doc/LAYERS.org

```diff
@@ -0,0 +1,540 @@
+Spacemacs treats extensions differently is so that it can do this.
+
+** packages.el
+This file should define a variable called =<layer>-packages=, which should be a
+list of all the packages that this layer needs. Additionally, a layer may
+/exclude/ packages through the variable =<layer>-excluded-packages=. This will
+block packages (and, despite the name, extensions) from being installed, even if
+they are included by another layer.
+
+For each included package, you may define one or more of the following
+functions, which are called in order by Spacemacs to initialize the package.
+
+1. =<layer>/pre-init-<package>=
+2. =<layer>/init-<package>=
+3. =<layer>/post-init-<package>=
+
+It is the responsibility of these functions to load and configure the package in
+question. Spacemacs will do nothing other than download the package and place it
+in the load path for you.
+
+*Note:* A package will not be installed unless at least one layer defines an
+=init= function for it. That is to say, in a certain sense, the =init= function
+does mandatory setup while the =pre-init= and =post-init= functions do optional
+setup. This can be used for managing cross-layer dependencies, which we will
+discuss later.
+
+** extensions.el
+The anatomy of =extensions.el= is mostly the same as that of =packages=. The
+extensions that the layer uses should be listed in =<layer>-pre-extensions=,
+which are those that will be initialized before the packages, and
+=<layer>-post-extensions=, which will be initialized after.
+
+As for packages, each extension should have one or more of the following
+functions defined. They work the same way as for packages.
+
+1. =<layer>/pre-init-<extension>=
+2. =<layer>/init-<extension>=
+3. =<layer>/post-init-<extension>=
+
+** config.el, funcs.el and keybindings.el
+These files contain no magic variables or functions, merely code that is
+executed. The intended use cases are
+
+- config.el :: configuration code that is independent of extensions and
+               packages, declaration layer variables, and so on
+- funcs.el :: definition of general utility functions
+- keybindings.el :: general keybindings
+
+The word /general/ here means /independent of any package or extension/. Since
+the end user can exclude an arbitrary set of packages, you cannot be sure that,
+just because your layer includes a package, that package will necessarily be
+loaded. For this reason, code in these files must be generally safe, regardless
+of which packages are installed.
+
+More on this in the next section.
+
+* The Spacemacs loading process
+The Spacemacs loading process can be summarized as follows:
+
+1. A handful packages are installed and loaded as part of the bootstrap process.
+   Among these packages, of particular interest to us, is =use-package=. All
+   subsequent code in the loading process may make use of it.
+2. Spacemacs goes through all the enabled layers and evaluates their files
```

Comment:
goes through all used layers in order of declaration. So it is possible to use this property to override a `configure` function and steal the ownership of a package in a custom layer.